chassis/core/src/main/java/com/griddynamics/jagger/master/MasterToJaasCoordinator.java

```diff
@@ -0,0 +1,307 @@
+        testEnvironmentEntity.setTestSuites(availableConfigs.stream()
+                                                            .map(TestSuiteEntity::new)
+                                                            .collect(Collectors.toList())
+        );
+        
+        return testEnvironmentEntity;
+    }
+    
+    private boolean isEnvIdUnacceptable(HttpClientErrorException exception) {
+        // If env with such an id already exists
+        if (exception.getStatusCode() == HttpStatus.CONFLICT) {
+            return true;
+        }
+        // If env id doesn't match acceptable pattern.
+        if (exception.getStatusCode() == HttpStatus.BAD_REQUEST && exception.getResponseBodyAsString().contains("doesn't match pattern")) {
+            return true;
+        }
+        return false;
+    }
+    
+    
+    private String extractSessionCookie(ResponseEntity<String> responseEntity) {
+        List<String> sessionCookies = responseEntity.getHeaders().get(HttpHeaders.SET_COOKIE).stream()
+                                                    .filter(s -> s.contains(TestEnvUtils.SESSION_COOKIE))
+                                                    .collect(Collectors.toList());
+    
+        if (sessionCookies.size() != 1) {
+            throw new IllegalStateException(String.format("There are %s values for '%s' cookie", sessionCookies.size(),
+                    TestEnvUtils.SESSION_COOKIE
+            ));
+        }
+    
+        String rawSessionCookie = sessionCookies.get(0);
+        int endIndex = rawSessionCookie.indexOf(";");
+        if (endIndex < 0) {
+            endIndex = rawSessionCookie.length() - 1;
+        }
+        return rawSessionCookie.substring(0, endIndex);
+    }
+    
+    public boolean isStandBy() {
+        return standBy;
+    }
+    
+    public boolean isRegistered() {
+        return registered;
+    }
+    
+    private final class StatusExchangeThread extends Thread {
+        
+        private String sessionCookie;
+    
+        private final SynchronousQueue<String> nextConfigToExecute = new SynchronousQueue<>();
+        
+        private volatile RequestEntity<TestEnvironmentEntity> requestEntity;
+    
+        public StatusExchangeThread(String sessionCookie) {
+            this.sessionCookie = sessionCookie;
+            this.requestEntity = buildPendingRequestEntity();
+
+            setName(this.getClass().getName());
+        }
+        
+        private RequestEntity<TestEnvironmentEntity> buildPendingRequestEntity() {
+            TestEnvironmentEntity testEnvEntity = buildTestEnvironmentEntityWith(TestEnvironmentStatus.PENDING);
+            HttpHeaders httpHeaders = new HttpHeaders();
+            httpHeaders.add(HttpHeaders.COOKIE, sessionCookie);
+            URI updatesUri = UriComponentsBuilder.newInstance()
+                                                 .uri(jaasEndpoint)
+                                                 .path("/envs")
```

Comment:
you can't give meaningful name to such a constant shorter then "/envs" so it will be more verbose